### PR TITLE
[JENKINS-32526] - Handle null @AncestorInPath when selecting Jobs/Dow…

### PR DESCRIPTION
**Pull request by Dohbedoh**
…nstream/Upstream Jobs and for autocompletion.

[JENKINS-32526](https://issues.jenkins-ci.org/browse/JENKINS-32526)

### CODE REVIEW

**Comment by Dohbedoh** [comment]
@reviewbybees

File: src/main/java/hudson/plugins/copyartifact/CopyArtifact.java [hunk 1]
```diff
@@ -577,7 +577,9 @@ public String resolve(String name) {
     public static final class DescriptorImpl extends BuildStepDescriptor<Builder> {
 
         public FormValidation doCheckProjectName(
-                @AncestorInPath Job<?,?> anc, @QueryParameter String value) {
+                @AncestorInPath AbstractProject anc, @QueryParameter String value) {
```
Comment (ikedam) [inline src/main/java/hudson/plugins/copyartifact/CopyArtifact.java]: Why do you need have it `AbstractProject`?
Pipeline (aka. workflow)'s snippet generator requires `Job`.
f28c55475678492a5e0659cb444b062fa5fe748b

File: src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java [hunk 2]
```diff
@@ -31,19 +31,10 @@
...
-import hudson.model.Cause;
-import hudson.model.FreeStyleProject;
-import hudson.model.FreeStyleBuild;
-import hudson.model.Item;
-import hudson.model.ParametersAction;
-import hudson.model.ParametersDefinitionProperty;
-import hudson.model.StringParameterDefinition;
-import hudson.model.StringParameterValue;
-import hudson.model.User;
-import hudson.model.Result;
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java]: I don't like wildcard import. (especially for classes)
Do you actually need this change?

File: src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java [hunk 3]
```diff
@@ -605,14 +614,75 @@ public void testCheckUpstreamBuildNumber() throws Exception {
...
+
+    private void testAutoCompleteUpstreamProjectName(
+            String [] expectedValues,
+            String value,
+            AbstractProject project,
+            DownstreamBuildSelector.DescriptorImpl d) {
+
+        assertArrayEquals(expectedValues, d.doAutoCompleteUpstreamProjectName(value, project).getValues().toArray());
+        //JENKINS-32526
+        assertArrayEquals(expectedValues, d.doAutoCompleteUpstreamProjectName(value, null).getValues().toArray());
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java]: It's better to sort the list or wrap it with `Set`.
`AutoCompleteCandidates` doesn't look ensure the order of values (strange behavior!)

File: src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java (hunk 2 above)
Comment (Dohbedoh) [inline src/test/java/hudson/plugins/copyartifact/DownstreamBuildSelectorTest.java]: No sorry. That's the IDE that organized import automatically.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 4]
```diff
@@ -188,6 +188,11 @@ public FormValidation doCheckUpstreamProjectName(
             if (containsVariable(upstreamProjectName)) {
                 return FormValidation.ok();
             }
+
+            if (project == null) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: While you are it, you should replace `AbstractProject` with `Job` to get Pipeline compatibility missed in #55. Ditto in other related validation methods. CC @apemberton: currently with Pipeline this throws a `NullPointerException`; fixing the PR as written would make it unconditionally return OK; also fixing the type of the parameter would make it actually perform the desired validation.

Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: [JENKINS-33578](https://issues.jenkins-ci.org/browse/JENKINS-33578) specifically.

**Comment by Dohbedoh** [comment]
@jglick Changed the variable type to `Job`. I needed to workaround a call to `AbstractProject#getRootProject()` in the `doCheckUpstreamProjectName` and `doCheckUpstreamBuildNumber` functions.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 5]
```diff
@@ -194,9 +194,16 @@ public FormValidation doCheckUpstreamProjectName(
                 // Jenkins is unavailable and validation is useless.
                 return FormValidation.ok();
             }
+
+            AbstractProject<?,?> upstreamRoot;
+            if(project != null && project instanceof AbstractProject) {
+                upstreamRoot = ((AbstractProject) project).getRootProject();
+            } else {
+                upstreamRoot = null;
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: No, you want to set it to `project`.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 6]
```diff
@@ -194,9 +194,16 @@ public FormValidation doCheckUpstreamProjectName(
...
+            AbstractProject<?,?> upstreamRoot;
+            if(project != null && project instanceof AbstractProject) {
+                upstreamRoot = ((AbstractProject) project).getRootProject();
+            } else {
+                upstreamRoot = null;
+            }
             
             AbstractProject<?,?> upstreamProject = jenkins.getItem(
-                    upstreamProjectName, project.getRootProject(), AbstractProject.class
+                    upstreamProjectName, upstreamRoot, AbstractProject.class
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: And here you want to deal with `Job`, not `AbstractProject`.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 7]
```diff
@@ -237,9 +244,16 @@ public FormValidation doCheckUpstreamBuildNumber(
                 // Jenkins is unavailable and validation is useless.
                 return FormValidation.ok();
             }
+
+            AbstractProject<?,?> rootProject;
+            if(project != null && project instanceof AbstractProject) {
+                rootProject = ((AbstractProject) project).getRootProject();
+            } else {
+                rootProject = null;
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: ditto

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 8]
```diff
@@ -237,9 +244,16 @@ public FormValidation doCheckUpstreamBuildNumber(
...
+            AbstractProject<?,?> rootProject;
+            if(project != null && project instanceof AbstractProject) {
+                rootProject = ((AbstractProject) project).getRootProject();
+            } else {
+                rootProject = null;
+            }
             
             AbstractProject<?,?> upstreamProject = jenkins.getItem(
-                    upstreamProjectName, project.getRootProject(), AbstractProject.class
+                    upstreamProjectName, rootProject, AbstractProject.class
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: ditto

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 9]
```diff
@@ -292,7 +306,9 @@ public AutoCompletionCandidates doAutoCompleteUpstreamProjectName(
                 @AncestorInPath AbstractProject<?,?> project
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: Need to make this `Job`.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 10]
```diff
@@ -292,7 +306,9 @@ public AutoCompletionCandidates doAutoCompleteUpstreamProjectName(
                 @AncestorInPath AbstractProject<?,?> project
         ) {
             // Specified Item to allow to autocomplete folders (maybe confusing...).
-            return AutoCompletionCandidates.ofJobNames(Item.class, value, project, project.getParent());
+            return project == null
+                    ? AutoCompletionCandidates.ofJobNames(Item.class, value, null)
```
Comment (jglick) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: If there is no context, just turn off completions (return `new AutoCompletionCandidates()`).

**Comment by jglick** [comment]
BTW if you want to _test_ Pipeline compatibility of completions, `WorkflowJob` is already in the test classpath. As I noted to @apemberton in [JENKINS-33577](https://issues.jenkins-ci.org/browse/JENKINS-33577), it is not really useful in this case anyway: `DownstreamBuildSelector` will not actually work.

File: src/main/resources/hudson/plugins/copyartifact/Messages.properties [hunk 11]
```diff
@@ -1,3 +1,4 @@
+CopyArtifact.AncestorIsNull=Ancestor/Context Unknown: the project specified cannot be validated
```
Comment (ikedam) [inline src/main/resources/hudson/plugins/copyartifact/Messages.properties]: I don't think this is helpful for users.
How about a text like "You look configuring non-buildable job (may be template) ..."?

Comment (jglick) [inline src/main/resources/hudson/plugins/copyartifact/Messages.properties]: Well it is not exactly “non-buildable”. But certainly the term “ancestor” will be meaningless to a user (this is an API term). I think it suffices to say that the configuration form is being displayed without enough context to determine which job will host the build step.

**Comment by ikedam** [comment]
You'd better add a comment when you add a new commit as github doesn't send a notification for commits.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 12]
```diff
@@ -196,7 +196,7 @@ public FormValidation doCheckUpstreamProjectName(
             }
             
             AbstractProject<?,?> upstreamProject = jenkins.getItem(
-                    upstreamProjectName, project.getRootProject(), AbstractProject.class
+                    upstreamProjectName, project, AbstractProject.class
```
Comment (ikedam) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: You might misunderstand the comment by jglick (https://github.com/jenkinsci/copyartifact-plugin/pull/79/files/65b30d6da1983e11be0f93d9defb664796fca99d#r56505288)
That meant you don't need to use `null` for `project` even when `project` doesn't have `getRootProject`.
`getRootProject` should be still required if it's available.

Though I don't know an actual case resulting `project.getRootProject() != project` in this context, it's better to implement in the same way as `DownstreamBuildSelector` works.

File: src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java [hunk 13]
```diff
@@ -177,7 +177,7 @@ protected boolean containsVariable(String str) {
          * @return
          */
         public FormValidation doCheckUpstreamProjectName(
-                @AncestorInPath AbstractProject<?,?> project,
+                @AncestorInPath Job<?,?> project,
```
Comment (ikedam) [inline src/main/java/hudson/plugins/copyartifact/DownstreamBuildSelector.java]: You should skip the validation if `project` is `null` (that is, when configuring a template).

**Comment by Dohbedoh** [comment]
@ikedam You're right I misunderstood jglick's point. I reimplemented the use of `getRootProject` and skip the validation in case `project` is `null`.
@reviewbybees

**Comment by ikedam** [comment]
Thanks for the update.
Looks good to me.

**Pull request merged**